Be able to count on iterator

plandolt opened the pull request:
This PR allows to count on `\Traversable` Objects. Not sure if I made the change to the correct branch. Tell me please if I should target 2.x or master.

plandolt committed:
Be able to count on iterator

fabpot commented on test/Twig/Tests/IntegrationTest.php:
```diff
@@ -307,3 +307,43 @@ public function __toString()
+/**
+ * This class is used in tests for the length filter
+ */
+class TraversableStub implements \Iterator
```
Instead of creating a class, could we use the `ArrayIterator` PHP built-in class instead?

plandolt commented:
@fabpot You were right. I updated the PR to just use php's built in ArrayIterator.

SpacePossum commented:
in the test it is better _not_ to use [ArrayIterator](http://php.net/manual/en/class.arrayiterator.php) as it implements `\Countable` so the unit test does not reach the new code (i.e. past `if ($thing instanceof \Countable || is_array($thing)) {`)

Looking at the PR itself one thing note:

```php
<?php
require_once __DIR__.'/vendor/autoload.php';

/**
 * This class is used in tests for the length filter
 */
class TraversableStub implements \Iterator
{
    private $position;
    private $data;

    public function __construct(array $data)
    {
        $this->position = 0;
        $this->data = $data;
    }

    public function current()
    {
        return $this->data[$this->position];
    }

    public function next()
    {
        ++$this->position;
    }

    public function key()
    {
        return $this->position;
    }

    public function valid()
    {
        return isset($this->data[$this->position]);
    }

    public function rewind()
    {
        $this->position = 0;
    }
}



$loader = new Twig_Loader_Array(array(
    'index' =>
    '{% for a in iterator %} {{ a }} {{ iterator|length }} {% endfor %}',
));
$twig = new Twig_Environment($loader);

$data = ['one', 'two', 'three'];

echo
    "\n------------\n".
    $twig->render('index', ['iterator' => new ArrayIterator($data)]).
    "\n------------\n".
    $twig->render('index', ['iterator' => $data]).
    "\n------------\n".
    $twig->render('index', ['iterator' => new TraversableStub($data)])
;
```

renders:

```
$ php test.php 

------------
 one 3  two 3  three 3 
------------
 one 3  two 3  three 3 
------------
 one 3 
```

This is the result of `iterator_count` modifying the current position of the iterator.
It is not (practical) possible to restore the position of the iterator, so this cannot be helped. Maybe something in the docs would be nice?

plandolt commented:
Good catch @SpacePossum!

I updated the PR and reintroduced the ``TraversableStub`` class to be used in the test. This will cover the new implementation. I also updated the doc with a warning section about the position might be changed (words from php.net) during counting on it. English is not my native language so please review that as well.

xabbuh commented:
Won't this break use cases where the provided `Traversable` implementation cannot be rewinded and someone tries to iterate the object after using the `length` filter?

nicolas-grekas commented:
> Won't this break use cases where the provided Traversable implementation cannot be rewinded and someone tries to iterate the object after using the length filter?

YES! the only safe interface is `IteratorAggregate`, so that's the one I'd recommend to use instead of `Traversable` in this PR.

plandolt commented:
@xabbuh yes it would break it. But also if you currently use ``length`` on a Traversable implementation you get ``1`` which is wrong too.

@nicolas-grekas && @xabbuh: I updated this PR to only use ``iterator_count`` on ``IteratorAggregate`` implementations. What about creating a PR to count on ``Iterator`` implementations (with the warning of not correct positions) or raising an exception if one tries to count on ``Iterator`` implementations in a next major?

fabpot commented on test/Twig/Tests/IntegrationTest.php:
```diff
@@ -307,3 +307,22 @@ public function __toString()
+    {
+        return new ArrayIterator($this->data);
+    }
+
```
This blank line should be removed according to our CS standards.

fabpot commented on test/Twig/Tests/Fixtures/filters/length.test:
```diff
@@ -16,6 +17,7 @@ return array(
     'number' => 1000,
     'to_string_able' => new ToStringStub('foobar'),
     'countable' => new CountableStub(42),       /* also asserts we do *not* call __toString() */
+    'traversable' => new IteratorAggregateStub(array('a', 'b', 'c')),   /* also asserts we do *not* call __toString() */
```
I would rename this variable to better reflect the class name.

nicolas-grekas commented:
> raising an exception if one tries to count on Iterator

I think we should follow PHP 7.2, which raises a deprecation notice when `count()` is used on non-countable types - then throw in Twig 3 yes.

plandolt commented:
@fabpot I updated the PR with a more accurate naming.

@nicolas-grekas I can create such a PR targeting the master branch in the future.

xabbuh commented on doc/filters/length.rst:
```diff
@@ -14,6 +14,8 @@ return value of the ``count()`` method.
 For objects that implement the ``__toString()`` magic method (and not ``Countable``),
 it will return the length of the string provided by that method.
 
+For objects that implement the ``IteratorAggregate`` interface, ``length`` will use the return value of the ``interator_count()`` method.
```
typo: `iterator_count()`

plandolt replied:
Thank you, fixed.

fabpot commented:
Thank you @scuben.
